model-serving/tfserving/README.md

```diff
@@ -0,0 +1,208 @@
+
+### Serving multiple models
+
+TensorFlow Serving is capable of serving multiple models on a single server. To enable this capability, you can specify a `model_config_file` as in the one provided in `config/multi_models.conf`. Please make sure that your saved models reside in `shared/models/`.
+
+1. Use the template in `config/multi_models.conf` and adapt it to your needs.
+
+2. Comment out the first block of code in `scripts/serving_entrypoint.sh` and uncomment the second block, so that the only uncommented lines are the following:
+
+```bash
+tensorflow_model_server --rest_api_port=8501 \
+       --model_config_file=/models/multi_models.conf \
+       "$@"
+```
+
+3. Re-build the serving image
+
+```bash
+$ docker build -t dlrs-serving:v0.2 -f Dockerfile.serving .
+```
+
+4. Run the container image again. You will be able to make requests to each of the models you uploaded by pointing to the model name.
+
+```bash
+$ docker run --rm -v ${PWD}/shared/models/:/models -p 8501:8501 dlrs-serving:v0.2
+
+# For model_a --> http://localhost:8501/v1/models/model_a/versions/1:predict 
+# For model_b --> http://localhost:8501/v1/models/model_b/versions/1:predict 
+
+# You should be able to see both models loaded:
+...
+2020-12-07 22:13:19.619699: I tensorflow_serving/core/loader_harness.cc:74] Loading servable version {name: model_a version: 1}
+2020-12-07 22:13:19.519714: I tensorflow_serving/core/loader_harness.cc:74] Loading servable version {name: model_b version: 1}
+2020-12-07 22:13:19.730794: I tensorflow_serving/model_servers/server.cc:387] Exporting HTTP/REST API at:localhost:8501
+```
+
+### Serving multiple versions
+
+Similarly to serving multiple models, you can serve multiple versions of a single model by creating a configuration file.
+
+1. Use the template in `config/multi_versions.conf` and adapt it to your needs.
+
+2. Comment out the first two blocks of code in `scripts/serving_entrypoint.sh` and uncomment the third block, so that the only uncommented lines are the following:
+
+```bash
+tensorflow_model_server --rest_api_port=8501 \
+       --model_name=${MODEL_NAME} --model_base_path=${MODEL_DIR} \
+       --model_config_file=/models/multi_versions.conf \
+       --allow_version_labels_for_unavailable_models=true \
+       "$@"
+```
+
+3. Re-build the serving image
+
+```bash
+$ docker build -t dlrs-serving:v0.3 -f Dockerfile.serving .
+```
+
+4. Run the container image again. You will be able to make requests to each of the models you uploaded by pointing to the model name.
+
+```bash
+$ docker run --rm -v ${PWD}/shared/models/:/models -p 8501:8501 dlrs-serving:v0.3
+
+# For my_model v1 --> http://localhost:8501/v1/models/my_model/versions/1:predict
+# Or by its label --> http://localhost:8501/v1/models/my_model/labels/stable:predict
+# For my_model v2 --> http://localhost:8501/v1/models/my_model/versions/2:predict
+# Or by its label --> http://localhost:8501/v1/models/my_model/labels/dev:predict
+```
+
+### Moniroting
```

Comment:
175 lines in and this is the first typo, very nice.

Reply:
Hahaha, oh my dyslexia